### ISSUE TITLE
Inconsistent Quadtree behavior with (zero height) items on the border

### ISSUE THREAD

**Issue by Rayne** [straker/kontra#111]
I'm currently in the process of adding the quadtree implementation to [my JS13k 2019 game](https://github.com/Rayne/js13k-lost-robot). My game utilizes line segments and I decided to add the bounding boxes of line segments to the quadtree. Sometimes, line segments have no height, e.g. the line segment from `(320,320)` to `(640,320)` has the bounding box `{x: 320, y:320, width: 320, height: 0}`. When the borders of a `Quadtree` instance align with such a bounding box, it is possible that the bounding box gets not returned when querying for objects with `Quadtree.get()`.

This behavior is not consistent. If the Quadtree isn't subdivided, then all entries will be returned. If the Quadtree is subdivided, then it is possible to loose items.

This could be a bug. One could also say that items on the border of `Quadtree.bounds` are not officially within bounds.

**Example** ([Source](https://gist.github.com/Rayne/b866164a8c550c36c80c2d1730163f8e))

```javascript
let quadtree = new kontra.Quadtree({
    bounds: {x: 0, y: 0, width: 100, height: 100}
});

let addExample = () => quadtree.add({x: 0, y: 100, width: 25, height: 0});
let test = query => console.log("Found " + quadtree.get(query).length + " items with " + JSON.stringify(query));

let queryA = {x: 0, y: 0, width: 50, height: 50};
let queryB = {x: 0, y: 99, width: 50, height: 50};
let queryC = {x: 0, y: 100, width: 50, height: 50};
let queryD = {x: 0, y: 101, width: 50, height: 50};

console.log("Adding three items to the quadtree");
for (let i = 0; i < 3; ++i) addExample();

test(queryA);
test(queryB);
test(queryC);
test(queryD);

console.log("Adding more items to subdivide the quadtree");
for (let i = 0; i < 97; ++i) addExample();

test(queryA);
test(queryB);
test(queryC);
test(queryD);
```

**Result**

```
          Adding three items to the quadtree quadtree.html:26:13
      OK: Found 3 items with {"x":0,"y":0,"width":50,"height":50} quadtree.html:19:33
      OK: Found 3 items with {"x":0,"y":99,"width":50,"height":50} quadtree.html:19:33
Maybe OK: Found 3 items with {"x":0,"y":100,"width":50,"height":50} quadtree.html:19:33
  Not OK: Found 3 items with {"x":0,"y":101,"width":50,"height":50} quadtree.html:19:33
          Adding more items to subdivide the quadtree quadtree.html:34:13
      OK: Found 0 items with {"x":0,"y":0,"width":50,"height":50} quadtree.html:19:33
  Not OK: Found 0 items with {"x":0,"y":99,"width":50,"height":50} quadtree.html:19:33
Maybe OK: Found 0 items with {"x":0,"y":100,"width":50,"height":50} quadtree.html:19:33
      OK: 0 items with {"x":0,"y":101,"width":50,"height":50}
```

**Expected**

- ```
    Not OK: Found 3 items with {"x":0,"y":101,"width":50,"height":50} quadtree.html:19:33
  ```

  Should not have returned items because the rectangle is out of bounds.
  On the other hand, returning too many items is a minor problem in this example

- ```
    Not OK: Found 0 items with {"x":0,"y":99,"width":50,"height":50} quadtree.html:19:33
  ```

  After adding items on the border, this should read `Found 100 items …` because

  ```
        OK: Found 3 items with {"x":0,"y":99,"width":50,"height":50} quadtree.html:19:33
  ```

  found all border items before subdividing the tree.

- Changing `bounds` to `bounds: {x: 0, y: 0, width: 100, height: 101}` is no valid workaround as it returns the same results. Switching to `bounds: {x: 0, y: 0, width: 100, height: 102}` works as expected

**Workaround**

Make `Quadtree.bounds.width` and `Quadtree.bounds.height` larger, e.g. `+2` per direction.

**Comment by straker** [comment]
Interesting. I'll have to take a look into this. I can kinda see how a 0 height/width element on the border could get lost.

**Comment by straker** [comment]
So I think the real issue is that any object that lies outside of the bounds of a split quadtree will not be added to the tree.

```js
let quadtree = new kontra.Quadtree({
  bounds: {x: 0, y: 0, width: 100, height: 100},
  maxObjects: 2 
});

quadtree.add({x: 105, y: 40, width: 10, height: 20});
quadtree.add({x: 105, y: 40, width: 10, height: 20});

console.log(quadtree._o.length)  // 2 (_o is the array of objects for each node)

// force the tree to split
quadtree.add({x: 105, y: 40, width: 10, height: 20});

console.log(quadtree._o.length)  // 0 (expected)
quadtree._s.forEach(node => console.log(node._o))  // all 4 subnodes (_s) have no objects in them
```

**Comment by straker** [comment]
Closed via #169
